contents/javascript/Javascript-modules.md

```diff
@@ -0,0 +1,204 @@
+
+    var getVariableOne = function() {
+        return variableOne;
+    }
+
+    var sumOfVariable = function() {
+        return variableOne + variableTwo;
+    }
+
+    var counter = function() {
+        variableOne++;
+    }
+
+    return {
+        getVariableOne: getVariableOne,
+        sumOfVariable: sumOfVariable,
+        counter,
+    }
+})()
+
+// index.js
+anExampleModule.getVariableOne(); // 1
+anExampleModule.sumOfVariable(); // 3
+```
+
+### CommonJS
+
+It is one of the most popular "unofficial" modules implementation because it is widely used by NodeJS. Nevertheless, it can be used without using NodeJS by using bundler <sup>[link](https://medium.com/@gimenete/how-javascript-bundlers-work-1fc0d0caf2da)</sup> such as Webpack <sup>[link](https://webpack.js.org/)</sup>.
+
+It uses the syntax `module.export` and `require`.
+
+`anExampleModule.js`
+```js
+var variableOne = 1;
+var variableTwo = 2;
+
+getVariableOne = function() {
+  return variableOne;
+}
+
+sumOfVariable = function() {
+    return variableOne + variableTwo;
+}
+
+module.exports = {
+  getVariableOne: getVariableOne,
+  sumOfVariable: sumOfVariable,
+};
+```
+
+`index.js`
+```js
+var anExampleModule = require('./anExampleModule.js');
+
+anExampleModule.getVariableOne(); // 1
+anExampleModule.sumOfVariable(); // 3
+```
+
+### EcmaScript2015 / ES6 modules
+
+In 2015, JavaScript introduced built-in modules in ES6. It uses the syntax `import` and `export`.
+
+`anExampleModule.js`
+```js
+var variableOne = 1;
+var variableTwo = 2;
+
+export getVariableOne() {
+    return variableOne;
+}
+
+export sumOfVariable() {
+    return variableOne + variableTwo;
+}
+```
+
+`index.js`
+```js
+import * as anExampleModule from './anExampleModule.js';
+
+anExampleModule.getVariableOne(); // 1
+anExampleModule.sumOfVariable(); // 3
+```
+
+### Which to use?
+
+While ES6 modules is officially the most recent JavaScript standard rolled out, it is still not widely supported by all browsers. It needs transpiler <sup>[link](https://scotch.io/tutorials/javascript-transpilers-what-they-are-why-we-need-them)</sup> such as Babel <sup>[link](https://babeljs.io/)</sup> for unsupported browser. Similar to CommonJS, ES6 modules also requires bundler such as Webpack.
+
+Despite ES6 modules not widely supported yet, it supports various advanced features not available in CommonJS such as asynchronous loading, support for tree shaking, static code analysis, etc. These features will not be covered for now.
+
+Nevertheless, both ES6 and CommonJS are a very valid approaches in modularizing your JavaScript project. You can take this opportunity to learn transpiling and bundling of JavaScript.
+
+The invention of ES6 and CommonJS has not made the old module pattern obsolete. It is still a valid approach in modularizing your code. It has the advantage of not requiring transpiling and bundling of your JavaScript code.
+
+All in all, the best approach depends on your project requirements.
+
+## How to start?
```

Comment:
Would some links to tutorials be useful in this section? They can explain **what** is happening and **how** it works in depth.